pkg/apis/crd/v1alpha1/types.go

```diff
@@ -600,3 +605,18 @@ type NamespacedName struct {
 	Name      string `json:"name,omitempty"`
 	Namespace string `json:"namespace,omitempty"`
 }
+
+// NetworkPolicyProtocol includes all protocols that `ports` can't support. All
+// fields should be used as a stand-alone field. To match all traffic with a
+// specific protocol, set the value of the corresponding field as an empty struct.
```

Comment:
I feel this can be decided per protocol. Maybe no need to define it with NetworkPolicyProtocol.

Reply:
If I understand you correctly, does the design you mean is the same as what I brought up before like this?
```
ingress/egress:
  type: array
  items:
    type: object
    required:
      - action
    properties:
      ...
      icmp:
        type: object
        properties:
          matchAll:
            type: boolean
          icmpSpecs:
            type: array
            items:
              type: object
              properties:
                icmpType:
                  type: integer
                  minimum: 0
                  maximum: 255
                icmpCode:
                  type: integer
                  minimum: 0
                  maximum: 255
```
I'm ok with both designs. Just think with `NetworkPolicyProtocol` it will be easy to add other protocol support like IGMP in the future. We don't have to add all protocols right under the `Rule`.

Reply:
No, I am not asking to change your current design. I meant this comment:

> To match all traffic with a specific protocol, set the value of the corresponding field as an empty struct.

needs not to be added on NetworkPolicyProtocol, as it sounds like protocol specific behavior, and maybe not the same for all protocols.

Reply:
Ah got it.